first draft of input mutation handling for aot autograd

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89532
* __->__ #88817



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

bdhirsh committed:
first draft of input mutation handling for aot autograd

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -310,7 +317,7 @@ def f(a):
         def f(a, b, c):
-            b.add_(1)
+            a.add_(1)
             c.add_(1)
```
Yeah, your problem is you're testing input mutation with addition. d/dx (x + 1) = 1; you get the same gradient no matter what. Change these to `mul_(2)` and you should start seeing numeric failures.

bdhirsh replied:
Ah yep - thanks. FWIW as of my changes last night, I think I now handle this case correctly (just changed to `.mul_(2)` locally and the tests still pass). The PR still isn't ready for review though, I'll give a shout when it is.

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
holler when to review

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh commented:
@ezyang should be ready for review now - CI failures aren't finished yet, but I tried to clean up most of my hacks and add a bunch of comments in `aot_autograd.py`.

Between your google doc and the code comments that I added, I think everything is pretty coverage - but let me know if you want me to go over stuff in more detail. I'll also add a better PR description later.

Right now I have the PR keeping the analysis pass around (and modified slightly so that "simple" input mutations go to aot autograd), but I'm leaning towards removing it completely, and making the error conditions in aot autograd clearer / louder. Most of them should be already.

bdhirsh commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -405,12 +405,27 @@ static PyObject* THPVariable__to_functional_tensor(
+      wrapped.set_requires_grad(self_.requires_grad());
+      if (wrapped.requires_grad()) {
+        impl::get_autograd_meta(wrapped)->grad_fn_ =
+            inner_autograd_meta->grad_fn_;
```
TODO: this is dangerous, and I'm going to change it to an "error on invocation grad fn". I think that exists somewhere in the code base, and just need to find it.

The idea is that if we have inputs to aot autograd that require gradients, that might have their own grad fn, then they will be marked with `input.is_leaf == False`, and we want the same to be true for our FunctionalWrapper inputs that we trace with.

However, the grad fn that we put on the input tensor wrappers should never be accessed. When we trace through the backward to create our joint graph, we specify the graph inputs as the primals, so autograd should stop there. Making the grad fn here an "error function" should assert that.

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+                else:
+                    updated_inpt = updated_inputs[curr_mutated_inpt_idx]
+                    curr_mutated_inpt_idx += 1
+                    # TODO: handle resize_() on inputs to a larger size.
```
This will be silent correctness if I turn off the analysis pass today, so I'll either fix resize_() or graph break in dynamo before landing.

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -55,6 +58,8 @@ def dynamo_timed(x):
 aten = torch.ops.aten
 
 
+KNOWN_TYPES = [torch.Tensor, int, str, float, bool, torch.SymInt, torch.SymFloat]
```
you could use `[...] + py_sym_types` to be more future-proofed 

(https://github.com/pytorch/pytorch/blob/0d87a4fec89fc78e568224935897ec585a6368a6/torch/fx/experimental/proxy_tensor.py#L59)

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
+# https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
+@dataclass(frozen=True)
+class ViewAndMutationMeta:
+    # length: # forward inputs
```
clean up?

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
+    # For every input, tells us whether the input:
+    # (a) is not mutated
+    # (b) only metadata is mutated
+    # (c) data (and maybe metadta) is mutated
```
typo

ezyang replied:
It's not clear to me why data mutation should subsume metadata mutation. They're pretty different beasts and it seems reasonable to track them separately. I'll be keeping an eye on this.

ezyang replied:
In particular, with this scheme you are obligated to update metadata whenever there is a data mutation, but that might be wasted work if the metadata didn't change

bdhirsh replied:
(very long delay - I have a large local refactor of AOT Autograd that does a bunch of cleanup, going to finally respond to PR feedback now)

This is true - the way this is handled on master now is in the MutationType.data case, we do a check at runtime to see if the metadata is different. I don't think this is too expensive (we're comparing sizes + strides + storage_offset), but agreed that precomputing a 4th enum would be cheaper.

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
+    # that way we can use it to replay the metadata mutation
+    metadata_mutation_input_info: List[Tensor]
+    # length: # outputs in the compiled forward. Equal to:
+    # length: (# inputs w data mutations) + (# outputs that don't alias inputs)
```
meant to put second `length:` here? 

also for readability maybe whitespace after var name before first comment so it's clear that length: goes with the thing below not above?  maybe even put the #length below the description and above the variable so it's sandwiched

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+            # This will be more complicated when you have multiple _base tensors aliasing the same
+            # underlying storage, when we eventually handle that.
+            # We'll need to ensure that we generate the view off of the right base.
+            inp_storage_refs = {StorageWeakRef(inpt.storage()): idx for idx, inpt in enumerate(flat_f_args)}
```
how come no check for is tensor on inp_storage_refs?  you can't call inpt.storage() on a non tensor right?

bdhirsh replied:
yep you're right

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                    # Then returning the output in the graph won't preserve the ._is_view() property,
+                    # because we'll be returning the "updated input" (which was created from an out-of-place op)
+                    if out_storage_ref in inp_storage_refs:
+                        maybe_aliased_inp_idx = inp_storage_refs[out_storage_ref]
```
what does 'maybe' refer to? is it uncertain that aliasing occured?  or just not certain whether one or more inputs were aliased (due to the comment about _base above)

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+    def maybe_to_fresh_input(idx, t):
+        if isinstance(t, Tensor) and meta.mutated_input_info[idx] == MutationType.data:
+            # Make sure the primal we pass to autograd.grad()
+            # seees the tensor before the mutation
```
typo (and one more of the same below)

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -261,11 +525,39 @@ def joint_forward_backward(
+        torch._enable_functionalization(reapply_views=True)
+        try:
+            # Run the joint
+            outs = joint_forward_backward(f_primals, f_tangents)
```
i assume this is just bc backwards may use inplace/mutation ops?  any other reason?

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -401,6 +693,179 @@ def disable_autocast_manager():
     finally:
         del guard
 
+def are_differentiable_views(view1, view2):
```
this could use a comment

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -457,52 +922,85 @@ def add_dupe_args(args):
-        _num_outs = len(out)
-    else:
-        _num_outs = 1
+    # This code only executes if we have graph inputs that alias each other, and one of those inputs
```
"This code only executes if..." refers to what, a hot path inside merge_view_inputs?

bdhirsh replied:
Updating to make this clearer - this code is run again at runtime, inside of `autograd.Function.forward()`, to create synthetic bases at runtime that we feed to the forward graph.

We only need to create synthetic bases though when we have:
(1) (at least) two graph inputs that are aliased to each other
(2) at least one of those graph inputs is mutated in the graph

And we precompute whether it's necessary at compile time. So we only do the "create synthetic base" step at runtime / in the hot path based on the precomputed info.

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -457,52 +922,85 @@ def add_dupe_args(args):
 
     disable_amp = torch._C._is_any_autocast_enabled()
 
     if config.use_functionalize:
```
do we want to kill this config (perhaps in a follow up PR..)?  or is the config going to live on but its new purpose is whether to run with decompositions?

ezyang replied:
yeah just kill the config

bdhirsh replied:
agreed, killing in the followup

wconstab commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+                    else:
+                        # TODO: when the aot autograd cache lands, we'll need to change this
+                        # to evaluate the sympy expressions given the inputs?
+                        size_ = [guard_int(x) for x in out_tensor_meta.size()]
```
is it worth passing the file/line into guard_int for traceability?

bdhirsh replied:
that would be a good point - fortunately, I'm killing all of this in the refactor :)

ezyang commented:
Need more PR description

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
+    # length: (# inputs that have *only* metadata mutations)
+    # We stash the updated FakeTensor that we traced with in the forward in here,
+    # that way we can use it to replay the metadata mutation
+    metadata_mutation_input_info: List[Tensor]
```
if this is guaranteed to be fake tensor, annotate it FakeTensor

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
+    #  idx of the input that this output aliases,
+    #   the FakeTensor representing that output, which we use to generate an as_strided() call with that metadata
+    # )
+    aliased_output_info: List[Optional[Union[int, torch.Tensor]]]
```
s/Union/Tuple/

is mypy running on this file 🤔

bdhirsh replied:
mypy is not running on this file :( it probably should...

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
 # TODO: Provide a faster version of this that assumes flat arguments
 # (so no pytree necessary)
-def detach_and_functionalize_pure(f, preserve_requires_grad=True):
+def run_functionalized_fw_and_collect_metadata(f):
```
sure would be useful to know what the return type of this function is

bdhirsh replied:
agreed, adding an annotation :)

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
-                return t
+    def inner(*args):
+        # This function is meant to be run with the forward, which expects a flat list of tensor/symint/other args.
+        assert all(isinstance(a, torch.Tensor) or type(a) in KNOWN_TYPES for a in args)
```
Are you sure you didn't want an `isinstance(a, KNOWN_TYPES)` test, where `KNOWN_TYPES` must be a tuple

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                # def compiled_fn(x):
+                #    out = traced_forward(x)
+                #    _x_updated_metadata = CompiledFunction.fw_metadata.metadata_mutation_input_info[0]
+                #    x.as_strided_(_x_updated_metadata.size(), _x_updated_metadata.stride(), _x_updated_metadata.storage_offset())
```
Does this happen inside the custom autograd function or not? Hard to tell from the example here

bdhirsh replied:
I'm going to make a fresh example along with my refactor. But in this example, the `.as_strided_()` call(s) happen outside of the custom autograd function.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                    inputs_with_mutated_data.append(new_arg)
+                    # For every mutated input, we ALSO need to return info on
+                    # whether than mutated input requires gradients. Why?
+                    # Our custom autograd.Function.forward returns updated inputs as outputs,
```
missing something here

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                    # We can use the storage aliasing of the inputs and updated inputs
+                    # to detect when an input was actually updated, or just inplace-viewed.
+                    collect_mutated_input_info.append(MutationType.metadata_only)
+                    collect_metadata_mutation_input_info.append(arg)
```
Whenever you're pushing onto multiple lists like this, it suggests that maybe you should have had a single list, with a dataclass that stores all the relevant info. This reduces the number of lists running around and reduces indexing confusion.

ezyang replied:
In this case, `collect_metadata_mutation_input_info` isn't pushed when it's not metadata mutation, I guess I'll need to see if it's helpful to have these split or not later.

bdhirsh replied:
Yep agreed - the two parallel lists here are totally unnecessary (mostly an artifact of the rushed merge to master)

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+            # Let's say inpA and inpB alias, and the user generated an output using out = inpA.view(...)
+            # For now, since we're not handling the case with multiple _base's sharing a storage,
+            # it is actually fine to arbitrarily pick which input to regenerate the aliased output from.
+            # e.g. out_new = inpB.as_strided(out.size(), out.stride(), out.storage_offset())
```
...are you sure? What if inpA requires grad, but inpB doesn't?

bdhirsh replied:
ugh, you're totally right. Adding a test for this (I'll confirm that the test fails before fixing)

bdhirsh replied:
So, the requires_grad example is actually fine, because we track requires-gradness separately (directly using the metadata we got from `run_functionalized_fw_and_collect_metadata`). But I'll add a test where it should actually break, e.g. if `inpA` and `inpB` have different sizes

bdhirsh replied:
Oh you know what... this actually works fine in the case of inputs that are aliased to each other.

I'm creating the set of input tensor storages using `flat_f_args`, which is the input arguments *after* converting them all to functional tensors. At that point, they all have unique storages, so we never have to worry about clobbering.

The "merge aliased inputs into synthetic bases" happens transparently after this step/

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
         torch._enable_functionalization(reapply_views=True)
         try:
-            outs = f(*f_args, **f_kwargs)
+            outs = f(*f_args)
```
Call this `f_outs` not outs. I don't know why I didn't name it like that originally.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                    # - the input was mutated and returned
+                    # - the input was a view (._is_view() is true)
+                    # Then returning the output in the graph won't preserve the ._is_view() property,
+                    # because we'll be returning the "updated input" (which was created from an out-of-place op)
```
This'll be clearer with an example

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                # Also, track the metadata to properly regenerate these outputs later.
+                # We store (inp_idx, fake_tensor_representing_output)
+                # That way later we can regenerate the output with inputs[inp_idx].as_strided(fake_tensor_output_metadata)
+                if maybe_aliased_inp_idx is None:
```
Merge this condition with above?

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+            requires_grad_out_info=collect_requires_grad_out_info,
+            aliased_output_info=collect_aliased_output_info,
+        )
+        return metadata, pytree.tree_map(from_fun, mutated_inps_and_outs)
```
NB: calling convention change was avoided by computing metadata in the preflight. This seems fine.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+    meta: ViewAndMutationMeta,
+    synthetic_base_info: List[Union[int, Tuple[int, List[Any]]]],
+):
+    def maybe_to_fresh_input(idx, t):
```
a quick top level comment here would help. I figured it out by reading this function in opposite direction (starting with bottom and working up)

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+            # seees the tensor before the metadata mutation
+            out = t.view(t.shape)
+        else:
+            out = t
```
nit just directly return instead of assigning to out

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+            return primals
+
+        f_args_inner = []
+        for outer_idx_or_lambda in synthetic_base_info:
```
what i wouldn't give for pattern matching here lol

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -401,6 +693,179 @@ def disable_autocast_manager():
+    #   idx corresponds to which synthetic base from the outer calling context to view
+    inner_calling_convention_meta: Dict[int, Union[int, Tuple[int, List[Any]]]] = {}
+    for aliased_input_indices in storage_ref_to_idx.values():
+        if len(aliased_input_indices) > 1 and any(
```
I'd suggest flipping the conditional, and then continue'ing in the simple case. Saves you an indent.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -401,6 +693,179 @@ def disable_autocast_manager():
+                # Regenerating views when reinterpreting complex / real tensors seems non-trivial,
+                # not handling for now
+                assert same_dtype_views(view1, view2), \
+                    "aot_autograd() does not yet handle input mutations on views with different dtypes."
```
The "we only check (a, b), (b, c), (c, d)" is kind of weird. Are these relations transitive?

bdhirsh replied:
yep, I think they are transitive (whether `v1` and `v2` have the same dtype, and share a differentiable base). I'll add a comment clarifying.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -401,6 +693,179 @@ def disable_autocast_manager():
+            if len(non_none_bases) == 0:
+                # Case where none of the aliases require gradients
+                example_idx = aliased_input_indices[0]
+                synthetic_base = torch.Tensor(fwd_inputs[example_idx].storage())
```
this will raise warning, use `_typed_storage()`, or better yet explicitly read out the type form tensor. cc @kurtamohler

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -457,52 +922,85 @@ def add_dupe_args(args):
-            lambda x: isinstance(x, Tensor) and not x.requires_grad, out
-        )
-    )
+    if isinstance(out, (list, tuple)):
```
can we make out always return a list?

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -513,41 +1011,81 @@ def add_dupe_args(args):
+            ]
+            fw_out_ids_requiring_grad = [
+                id(x) for (i, x) in enumerate(fw_outs[:num_forward_returns])
+                if isinstance(x, torch.Tensor) and CompiledFunction.fw_metadata.requires_grad_out_info[i]
```
This variable looks dead, what's going on

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+            # Generate: the updated args, including (potentially multiple) synthetic bases
+            # that replace the views. The input views are regenerated manually in the compiled function.
+            # TODO: think harder about what happens if (a view of) one of these mutated input views is ALSO returned
+            new_inputs, metadata = merge_view_inputs(no_dupe_args, CompiledFunction.fw_metadata.mutated_input_info)
```
So `merge_view_inputs` is hotpath. We may need to more carefully audit it for speed

bdhirsh replied:
Agreed - I didn't try to carefully optimize it for now just because I think the "mutate an input that aliases another input" is pretty uncommon, but this is probably worth coming back to.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+            curr_metadata_only_mutated_inpt_idx = 0
+            for inpt_idx, mutation_type in enumerate(CompiledFunction.fw_metadata.mutated_input_info):
+                if mutation_type == MutationType.none:
+                    continue
```
For hotpath speed, it would be better to enumerate over indices where work is known to be needed to be done, rather than having to skip over inputs that are noop

bdhirsh replied:
fair - I'll precompute it.

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+                    # and updated_inpt is always densely packed.
+                    if original_inpt.size() != updated_inpt.size() \
+                            or original_inpt.stride() != updated_inpt.stride() \
+                            or original_inpt.storage_offset() != updated_inpt.storage_offset():
```
I'm surprised this check can't be precomputed

ezyang replied:
oh this is probably thing I mentioned about separating metadata and data mutation flags

bdhirsh replied:
yep - fixed in upcoming PR

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+        # compiling them.
+        if CompiledFunction.num_aliased_outputs > 0:
+            assert CompiledFunction.num_aliased_outputs + len(fw_outs) == len(CompiledFunction.fw_metadata.aliased_output_info)
+            fw_outs_including_aliases = []
```
it might be profitable to avoid allocating a new list, and instead modify the old list in place. idk, maybe it doesn't matter

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+                        fw_outs_including_aliases.append(input_alias)
+                    else:
+                        # TODO: when the aot autograd cache lands, we'll need to change this
+                        # to evaluate the sympy expressions given the inputs?
```
This TODO is pretty important

bdhirsh replied:
Mostly irrelevant in the new world where we return all aliases as outputs in the compiled graph

ezyang reviewed (approved):
This is really nice work! Good job!

wconstab commented on torch/_dynamo/variables/torch.py:
```diff
@@ -381,6 +381,25 @@ def get_state_from_generator():
+                    for x in args
+                ]
+            )
+            if any_symints_or_symfloats and all_ints_or_floats:
```
so this basically checks for a case where there are no tensor inputs?

also `all_ints_or_floats` technically means `all_ints_or_floats_or_symints_or_symfloats` right?  why is the `any` variable even necessary?

wait, i almost missed this- is there a diff between DynamicShapeVariable and variables.DynamicShapeVariable?

bdhirsh replied:
So this was wrong - I just updated this with a way more conservative version. But the idea behind this change was:

(1) Now that we're allowing graphs with input mutations into aot autograd, more graphs are going through our stack and (potentially) exposing more bugs

(2) One bug that showed up was from this code in `hf_Longformer`:

```
              hidden_states = hidden_states.view(
                  hidden_states.size(0),
                  torch.div(hidden_states.size(1), (window_overlap * 2), rounding_mode="trunc"),
                  window_overlap * 2,
                  hidden_states.size(2),
              )
```

Where we're calling torch.div(symint, int). `torch.div()` always creates a tensor, and we're passing that tensor as a size argument into a view call, which immediately calls `.item()` on the created tensor.

That breaks today, because fake tensor doesn't support const-propagating symints that are stored as tensor data. So I added this graph break as a dumb workaround for now.

wconstab reviewed (approved):
This looks great, and you made it quite readable so thanks for that!

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on functorch/_src/aot_autograd.py:
```diff
@@ -131,29 +137,91 @@ def posthook(grad_input, grad_output):
+    if aliased_base_tensor.is_complex() and not target_meta_tensor.is_complex():
+        aliased_out = torch.view_as_real(aliased_base_tensor).as_strided(size, stride, storage_offset)
+    elif not aliased_base_tensor.is_complex() and target_meta_tensor.is_complex():
+        aliased_out = torch.view_as_complex(aliased_base_tensor).as_strided(size, stride, storage_offset)
```
any view function that can fail needs to be infallible

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
I still need to go back and address a ton of feedback, but CI is green so I'm going to go ahead and land. This should fix the majority of input mutation issues with AOT Autograd (batch norm requires some extra fixes, currently pushing on that)

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "first draft of input mutation handling for aot autograd"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bdhirsh commented:
an input mutation test was failing locally for me on CPU when all inputs didn't require grad so I tried changing them to require gradients (so we run the new logic), but that test failed - worth looking into.

But now I'm thinking that my failure locally on CPU must have been spurious, so I changed the test back to the old behavior. Letting CI run on it

bdhirsh commented:
@pytorchbot merge

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+    fn,
+    *,
+    meta: ViewAndMutationMeta,
+    synthetic_base_info: List[Union[int, Tuple[int, List[Any]]]],
```
definitely is. I'm going to save this refactor for when I/we replace the as_strided code here with view replay (which we definitely wanna do so we get an easier-to-compile graph, although "synthetic_bases" is probably a relatively rare case).

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -261,11 +525,39 @@ def joint_forward_backward(
+        finally:
+            torch._disable_functionalization()
+
+        # Syncing of inputs/outputs was already done directly in the joint call
```
yep you're right - the syncing is redundant but it's easier to leave it in. I'll just remove the comment.

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                # instead of a general as_strided() call.
+                # This could also improve perf, since today this will cause
+                # more as_strided_scatter() ops in the graph.
+                view_arg = outer_base.as_strided(*strided_args)
```
agreed - alban exposed view-replay to python last week. I'll try to fix this in a followup

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+        for i, arg in enumerate(primals_):
+            if not isinstance(arg, Tensor):
+                continue
+            torch._sync(arg)
```
We need the tensors to by synced by the time we call `autograd.grad()` to trace the backwards- but there's nothing ensuring that we sync before then.

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -401,6 +693,179 @@ def disable_autocast_manager():
+            non_none_bases = [fwd_inputs[i]._base for i in aliased_input_indices if fwd_inputs[i]._base is not None]
+            aliases_with_none_bases = [fwd_inputs[i] for i in aliased_input_indices if fwd_inputs[i]._base is None]
+            if len(non_none_bases) == 0:
+                # Case where none of the aliases require gradients
```
Yep thanks - updating the comment to:
```
           # Case where none of the aliases have a ._base
           # we generate a synthetic base without gradients, and generate views off of it
```

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -513,41 +1011,81 @@ def add_dupe_args(args):
             fw_outs_not_requiring_grad = [
-                x for (i, x) in enumerate(fw_outs[0:num_outs]) if CompiledFunction.flat_outs_not_requiring_grad[i]
+                x for (i, x) in enumerate(fw_outs[:num_forward_returns])
+                if isinstance(x, torch.Tensor) and not CompiledFunction.fw_metadata.requires_grad_out_info[i]
```
yeah... we could remove it if you think it's non-negligible perf.

We need it because the way that I set up the metadata was that `fw_metadata.requires_grad_info` is a giant `List[bool]`, where every output of the compiled forward graph gets an entry.

It's only true for outputs that do require gradient, and false for everything else (aka tensors that don't require grad, and also non-tensors).

If we want, I could change that list so that True represents "this output is a tensor that does not require gradients" and False is everything else.

bdhirsh commented on torch/_dynamo/optimizations/analysis.py:
```diff
@@ -155,11 +155,17 @@ def _wrap_to_fake_tensor(t, *, f_mode):
+                    inps_that_alias_node = [
+                        n
+                        for n in inp_nodes
+                        if n.meta["alias_groups"] & node.meta["alias_groups"]
```
delayed reply, but I ended up removing my changes here - we no longer run this analysis pass at all. I agree that the meta info is pretty hard to understand.

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -243,14 +244,58 @@ def _outs_and_grads(fn, inps):
-            compiled_f = aot_function(f, nop)
-        ref_out, ref_grad = _outs_and_grads(f, inp)
-        test_out, test_grad = _outs_and_grads(compiled_f, inp)
-        self.assertEqual(ref_out, test_out)
```
bleh good point - fixing

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -259,6 +304,21 @@ def verify_aot_autograd(self, f, inp):
+            if isinstance(ref_i, torch.Tensor):
+                self.assertEqual(ref_i.requires_grad, test_i.requires_grad)
+                self.assertEqual(ref_i, test_i)
+        if return_fw_graph:
```
I thought it might be confusing to return something that 80% of our tests ignore, but you're probably right - that's less confusing than Yet Another Kwarg

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -243,14 +244,58 @@ def _outs_and_grads(fn, inps):
+            inp, graph_inps = inp_callable()
+        else:
+            inp_copy = []
+            # Our input clones need to mimic when inputs are duplicates of one another
```
late, but - yep, just to simplify some test boilerplate. If you think that's more confusing, I can change it.
